### PR TITLE
Fix ConnServer stall

### PR DESCRIPTION
**Pull request by luisremis**
This PR fixes an issue reported by user where the server will stop taking request, usually after a parallel workload failed or had to be paused.

#### Reproduce

The issue was reproduced by starting and stopping heavy parallel workload on a large database until the error occurred (usually after 2 and 4 tries)    

#### Diagnostic 

The single threaded ConnServer object does a blocking accept(), and after accepting the connection, the negotiation protocol is done. If the client stops responding in the middle of the protocol negotiation, the single threaded ConnServer will be blocked and stayed blocked until the client responds or the connection is closed.

### CODE REVIEW

File: src/comm/ConnServer.cc [hunk 1]
```diff
@@ -114,53 +121,72 @@ ConnServer::ConnServer(int port, ConnServerConfig config) :
...
         THROW_EXCEPTION(ProtocolError, "Protocol negotiation failed");
     }
 }
+
+std::unique_ptr<Connection> ConnServer::accept()
+{
+    auto connected_socket = TCPSocket::accept(_listening_socket);
+
+    auto tcp_connection = std::unique_ptr<TCPConnection>(
+        new TCPConnection(std::move(connected_socket), _config.metrics));
```
Comment (JoshStoddard) [inline src/comm/ConnServer.cc:189]:
```suggestion
    auto tcp_connection = std::make_unique<TCPConnection>(
        std::move(connected_socket), _config.metrics));
```

File: test/TCPConnectionTests.cc [hunk 2]
```diff
@@ -33,7 +33,8 @@ TEST(TCPConnectionTests, SyncMessages)
 
         barrier.wait();
 
-        auto server_conn = server.accept();
+        std::shared_ptr<comm::Connection> server_conn = server.accept();
+        server_conn = server.negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/TCPConnectionTests.cc:37]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/TCPConnectionTests.cc [hunk 3]
```diff
@@ -81,7 +82,8 @@ TEST(TCPConnectionTests, AsyncMessages)
...
```
Comment (JoshStoddard) [inline test/TCPConnectionTests.cc:86]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/TCPConnectionTests.cc [hunk 4]
```diff
@@ -130,7 +132,8 @@ TEST(TCPConnectionTests, ServerShutdownRecv)
 
         barrier.wait();
 
-        server.accept();
+        std::shared_ptr<comm::Connection> server_conn = server.accept();
+        server_conn = server.negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/TCPConnectionTests.cc:136]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/TCPConnectionTests.cc [hunk 5]
```diff
@@ -159,7 +162,8 @@ TEST(TCPConnectionTests, SendArrayInts)
 
         barrier.wait();
 
-        auto server_conn = server.accept();
+        std::shared_ptr<comm::Connection> server_conn = server.accept();
+        server_conn = server.negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/TCPConnectionTests.cc:166]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/TLSConnectionTests.cc [hunk 6]
```diff
@@ -50,7 +50,8 @@ TEST_F(TLSConnectionTests, SyncMessages)
 
         barrier.wait();
 
-        auto server_conn = server.accept();
+        std::shared_ptr<comm::Connection> server_conn = server.accept();
+        server_conn = server.negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/TLSConnectionTests.cc:54]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/TLSConnectionTests.cc [hunk 7]
```diff
@@ -98,7 +99,8 @@ TEST_F(TLSConnectionTests, AsyncMessages)
...
```
Comment (JoshStoddard) [inline test/TLSConnectionTests.cc:103]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/TLSConnectionTests.cc [hunk 8]
```diff
@@ -147,7 +149,8 @@ TEST_F(TLSConnectionTests, ServerShutdownRecv)
 
         barrier.wait();
 
-        server.accept();
+        std::shared_ptr<comm::Connection> server_conn = server.accept();
+        server_conn = server.negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/TLSConnectionTests.cc:153]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/TLSConnectionTests.cc [hunk 9]
```diff
@@ -176,7 +179,8 @@ TEST_F(TLSConnectionTests, SendArrayInts)
 
         barrier.wait();
 
-        auto server_conn = server.accept();
+        std::shared_ptr<comm::Connection> server_conn = server.accept();
+        server_conn = server.negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/TLSConnectionTests.cc:183]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/TLSConnectionTests.cc [hunk 10]
```diff
@@ -263,7 +267,8 @@ TEST_F(TLSConnectionTests, UseAutogeneratedCertificates)
 
         barrier.wait();
 
-        ASSERT_NO_THROW(server.accept());
+        std::shared_ptr<comm::Connection> server_conn = server.accept();
+        server_conn = server.negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/TLSConnectionTests.cc:271]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

File: test/VDMSServer.cc [hunk 11]
```diff
@@ -16,6 +16,7 @@ VDMSServer::VDMSServer(int port, comm::ConnServerConfig config) :
     auto thread_function = [&]()
     {
         std::shared_ptr<comm::Connection> server_conn = _server->accept();
+        server_conn = _server->negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/VDMSServer.cc:19]:
```suggestion
        auto server_conn = _server.negotiate_protocol(_server.accept());
```

File: test/AuthEnabledVDMSServer.cc [hunk 12]
```diff
@@ -39,6 +39,7 @@ AuthEnabledVDMSServer::AuthEnabledVDMSServer(int port, AuthEnabledVDMSServerConf
     auto thread_function = [&]()
     {
         std::shared_ptr<comm::Connection> server_conn = _server.accept();
+        server_conn = _server.negotiate_protocol(server_conn);
```
Comment (JoshStoddard) [inline test/AuthEnabledVDMSServer.cc:42]:
```suggestion
        auto server_conn = server.negotiate_protocol(server.accept());
```

**Review by JoshStoddard — DISMISSED** [review]

File: src/comm/ConnServer.cc [hunk 13]
```diff
@@ -47,6 +47,13 @@
 #include "comm/TLSConnection.h"
 #include "comm/Variables.h"
 
+#include "util/gcc_util.h"
+DISABLE_WARNING(effc++)
+DISABLE_WARNING(suggest-override)
+#include <glog/logging.h>
+ENABLE_WARNING(suggest-override)
+ENABLE_WARNING(effc++)
+
```
[the remark this answers is not in the document]
Comment (luisremis) [reply src/comm/ConnServer.cc:56]: not at all, thanks.

Comment (luisremis) [inline src/comm/ConnServer.cc:56]:
```suggestion
```

File: src/comm/ConnServer.cc [hunk 14]
```diff
@@ -114,53 +121,72 @@ ConnServer::ConnServer(int port, ConnServerConfig config) :
 
 ConnServer::~ConnServer() = default;
 
-std::unique_ptr<Connection> ConnServer::accept()
+// c contains a TCPConnection, unencrypted, connection to a client.
+// The ConnServer will implement the protocol negotiation.
+// This right now is a simple handshake, design for ApertureDB Server use-case.
+// This protocol can be a virtual method in the future to support arbitrary protocols.
+std::shared_ptr<Connection> ConnServer::negotiate_protocol(std::shared_ptr<Connection> conn)
```
[the remark this answers is not in the document]
Comment (luisremis) [reply src/comm/ConnServer.cc:128]: mmm yes. let me give it a try.

File: include/util/Utils.h [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+#pragma once
+
+#include <util/TypeName.h>
+#include <iostream>
+
+template<typename T>
+void print_binary_data(const T& data)
+{
+    // print size of hellomessage
```
Comment (luisremis) [inline include/util/Utils.h:15]:
```suggestion
```

File: src/comm/ConnServer.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (luisremis) [reply src/comm/ConnServer.cc]: partial success. it seems we can return unique_ptr, but if we take a unique_ptr as a parameter, we cannot use static_pointer_cast on it, so we have to use raw ptrs instead. not sure what is preferable. 

Here some interesting explanations I found:
https://groups.google.com/g/comp.std.c++/c/IOmEoO0KTg0?pli=1
https://stackoverflow.com/questions/36120424/alternatives-of-static-pointer-cast-for-unique-ptr

[the remark this answers is not in the document]
Comment (JoshStoddard) [reply src/comm/ConnServer.cc]: One approach might be to change the return type of `ConnServer::accept()` and the arg type of `ConnServer::negotiate_protocol()` to `std::unique_ptr< TCPConnection >`.

**Pull request merged**